plugins/spdk/spdk-nvme.c

```diff
@@ -0,0 +1,318 @@
+#include "nvme-print.h"
+#include "plugin.h"
+#include "nvme/tree.h"
+#include "nvme/private.h"
+
+#include "util/logging.h"
+#include "util/suffix.h"
+
+#define CREATE_CMD
+#include "spdk-nvme.h"
+
+static struct stat nvme_stat;
+static char *spdk_dev = "/dev/spdk";
+
+static char *output_format_val = "normal";
+
+#define NVME_ARGS(n, c, ...)                                                      \
+	struct argconfig_commandline_options n[] = {                              \
+		OPT_FLAG("verbose",      'v', NULL,               verbose),       \
+		OPT_FMT("output-format", 'o', &output_format_val, output_format), \
+		##__VA_ARGS__,                                                    \
+		OPT_END()                                                         \
+	}
+
+static bool is_chardev(void)
+{
+	return S_ISCHR(nvme_stat.st_mode);
+}
+
+static bool is_blkdev(void)
+{
+	return S_ISBLK(nvme_stat.st_mode);
+}
+
+static int spdk_open_dev(char *dev, int flags)
+{
+	int err, fd;
+
+	err = open(dev, flags);
+	if (err < 0)
+		goto perror;
+
+	fd = err;
+
+	err = fstat(fd, &nvme_stat);
+	if (err < 0) {
+		close(fd);
+		goto perror;
+	}
+
+	if (!is_chardev() && !is_blkdev()) {
+		fprintf(stderr, "%s is not a block or character device\n", dev);
+		close(fd);
+		return -ENODEV;
+	}
+
+	return fd;
+perror:
+	perror(dev);
+	return err;
+}
+
+static inline void spdk_nvme_id_ns_flbas_to_lbaf_inuse(__u8 flbas, __u8 *lbaf_inuse)
+{
+	*lbaf_inuse = (((flbas & NVME_NS_FLBAS_HIGHER_MASK) >> 1) |
+			(flbas & NVME_NS_FLBAS_LOWER_MASK));
+}
+
+static void _nvme_ns_set_names(struct nvme_ns *n, const char *name)
+{
+	char nname[PATH_MAX];
+
+	sprintf(nname, "%s/%s", spdk_dev, name);
+	n->generic_name = strdup(nname);
+	n->name = strdup(nname);
+}
+
+static int spdk_nvme_ns_init(struct nvme_ns *n)
+{
+	struct nvme_id_ns ns = { };
+	uint8_t flbas;
+	int ret;
+
+	ret = nvme_ns_identify(n, &ns);
+	if (ret)
+		return ret;
+
+	spdk_nvme_id_ns_flbas_to_lbaf_inuse(ns.flbas, &flbas);
+	n->lba_shift = ns.lbaf[flbas].ds;
+	n->lba_size = 1 << n->lba_shift;
+	n->lba_count = le64_to_cpu(ns.nsze);
+	n->lba_util = le64_to_cpu(ns.nuse);
+	n->meta_size = le16_to_cpu(ns.lbaf[flbas].ms);
+
+	return 0;
+}
+
+static nvme_ns_t nvme_ns_open(const char *name, char *path)
+{
+	struct nvme_ns *n;
+
+	n = calloc(1, sizeof(*n));
+	if (!n) {
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	_nvme_ns_set_names(n, name);
+	n->fd = spdk_open_dev(path, O_RDONLY);
+	if (n->fd < 0)
+		goto free_ns;
+
+	if (nvme_get_nsid(n->fd, &n->nsid) < 0)
+		goto close_fd;
+
+	if (spdk_nvme_ns_init(n) != 0)
+		goto close_fd;
+
+	list_head_init(&n->paths);
+	list_node_init(&n->entry);
+
+	return n;
+
+close_fd:
+	close(n->fd);
+free_ns:
+	free(n->name);
+	free(n);
+
+	return NULL;
+}
+
+// Since some data are not null terminated, trim trailing spaces
+void trim(char **p)
+{
+	if (!p)
+		return;
+
+	char *tmp = *p;
+
+	if (!tmp)
+		return;
+
+	for (int i = strlen(tmp) - 1; i >= 0; --i) {
+		if (tmp[i] !=  ' ') {
```

Comment:
`(tmp[i] !=  ' ')` is correct. Starts at the end of array and looking for first non-empty character, and then put `\0` at n+1.

Reply:
This function does an out bound access.

Reply:
Sorry, but I don't see it. I assumed that strings are null terminated and start from strlen(tmp) -1, to avoid out bound access.

Reply:
If I may butt in... ;-)

Here's another way to trim trailing spaces:
```
        tmp[strcspn(tmp, ' ')] = '\0';
```

Reply:
I think that ` tmp[strcspn(tmp, ' ')] = '\0';` may trim all characters after first space, not only trailing spaces.

Reply:
You are correct. I didn't realize that there might be spaces inside the string.

Reply:
I've extracted the function into a simple test and when the string is allocated on the stack the program crashed. Didn't really look into the details.

We have similar code the code base and there were fallouts because of out of bound accesses. Thus I am a bit cautious.